DTCOR-20 - iOS - Add content provider with FRC support

soxjke opened the pull request.

soxjke committed:
DTCOR-20 - iOS - Add content provider with FRC support

- Added content provider
- Added objects getter
- Added objects mutator

soxjke committed:
DTCOR-20 - iOS - Add content provider with FRC support

- Mutator implementation
- Section adapter

soxjke committed:
DTCOR-20 - iOS - Add content provider with FRC support

- Section adapter implementation

soxjke committed:
DTCOR-20 - iOS - Add content provider with FRC support

- Constant fetch request FRC support

soxjke committed:
DTCOR-20 - iOS - Add content provider with FRC support

- Mutation of fetch request support
- Added spec

soxjke committed:
Merge remote-tracking branch 'origin/develop' into DTCOR-20-add-content-provider-with-frc-support

Conflicts:
	DataSources.xcodeproj/project.pbxproj

soxjke committed:
DTCOR-20 - iOS - Add content provider with FRC support

- FRC content provider
- sample app with core data stack

soxjke committed:
DTCOR-20 - iOS - Add content provider with FRC support

- Added reordering support to fetched results controller content provider + sample

soxjke committed:
DTCOR-20 - iOS - Add content provider with FRC support

- Improved sample app
- Corrected ContentProvider implementations on insert/delete to correspond NSFetchedResultsControllerDelegate implementation
- Updated specs
- FRC workarounds

soxjke committed:
DTCOR-20 - iOS - Add content provider with FRC support

- Completed FRC workarounds

sergeyzenchenko commented on DataSources/CoreDataManager.m:
```diff
@@ -0,0 +1,183 @@
+        NSFetchRequest *frPerson = [NSFetchRequest fetchRequestWithEntityName:NSStringFromClass([Person class])];
+        NSFetchRequest *frPersonOrdered = [NSFetchRequest fetchRequestWithEntityName:NSStringFromClass([PersonOrdered class])];
+        NSArray *persons = [main executeFetchRequest:frPerson error:NULL];
+        if (!persons.count) {
```
countForFetchRequest maybe?

soxjke replied:
ok

sergeyzenchenko commented on DataSources/CoreDataManager.m:
```diff
@@ -0,0 +1,183 @@
+                [background save:NULL];
+            }];
+        }
+        NSArray *personsOrdered = [main executeFetchRequest:frPersonOrdered error:NULL];
```
countForFetchRequest maybe?

soxjke replied:
ok

soxjke committed:
Merge remote-tracking branch 'origin/DTCOR-20-add-content-provider-with-frc-support' into DTCOR-20-add-content-provider-with-frc-support

Conflicts:
	DataSources/TECFetchedResultsControllerContentProvider.m

soxjke committed:
DTCOR-20 - iOS - Add content provider with FRC support

- Fixed review notes

soxjke committed:
DTCOR-20 - iOS - Add content provider with FRC support

- Workarounds sample

soxjke committed:
DTCOR-20 - iOS - Add content provider with FRC support

- Sample up code cleanup

soxjke committed:
DTCOR-20 - iOS - Add content provider with FRC support

- Sample up workarounds screen

fuzza commented on DataSources/AppDelegate.m:
```diff
@@ -16,7 +17,7 @@ @implementation AppDelegate
 
 - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
-    // Override point for customization after application launch.
+    [[CoreDataManager sharedObject] setup];
```
I think we should use MagicalRecord. It will simplify network stack initializing and also will give us ability to use in-memory stack for unit testing

fuzza commented on DataSources/Base.lproj/Main.storyboard:
```diff
@@ -17,9 +139,136 @@
+    //     "ins row {0, 2}",
+    // Or even better just get rid of them
+    // 5. (Check with 2. First tap - 2, then - 5)
+    // Move from-to same indexpath should be replaced by update
```
It should be moved to documentation or code comment. Even better would be to have some UI tests suites for different iOS versions.

fuzza commented on DataSources/CoreDataManager.m:
```diff
@@ -0,0 +1,181 @@
+                    Person *person = [[Person alloc] initWithEntity:personEntityDescription insertIntoManagedObjectContext:background];
+                    person.name = name;
+                    person.phone = TEST_PERSON_ARRAY_PHONE[idx];
+                }];
```
too many indentation levels, few methods should be extracted from this block of code

fuzza commented on DataSources/CoreDataManager.m:
```diff
@@ -0,0 +1,181 @@
+#import "PersonOrdered.h"
+
+#define TEST_PERSON_ARRAY_NAME @[@"Alexey Fayzullov", @"Anastasiya Gorban", @"Petro Korienev", @"Sergey Zenchenko"]
+#define TEST_PERSON_ARRAY_PHONE @[@"+380670000000", @"+380670000001", @"+380670000002", @"+380670000003"]
```
using dictionary here will be much clearer

fuzza commented on DataSources/CoreDataManager.m:
```diff
@@ -0,0 +1,181 @@
+                    personOrdered.name = name;
+                    personOrdered.phone = TEST_PERSON_ARRAY_PHONE[idx];
+                    personOrdered.ordinal = @(idx);
+                }];
```
It seems this code also should be extracted to separate method

fuzza commented on DataSources/CoreDataManager.m:
```diff
@@ -0,0 +1,181 @@
+            NSArray *objects = [background executeFetchRequest:fetchRequest error:NULL];
+            for (NSManagedObject *object in objects) {
+                [background deleteObject:object];
+            }
```
extract to separate method

fuzza commented on DataSources/CoreDataManager.m:
```diff
@@ -0,0 +1,181 @@
+    return [NSURL fileURLWithPath:NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES).firstObject];
+}
+
+- (NSManagedObjectModel *)managedObjectModel {
```
Let's use MR for stack initialization. It will significantly simplify example code and will remove boilerplate code

fuzza commented on DataSources/Person+CoreDataProperties.h:
```diff
@@ -0,0 +1,23 @@
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface Person (CoreDataProperties)
```
I think we should use MOGenerator as it provides code-generated key-paths and other useful stuff

fuzza commented on DataSources/Person.m:
```diff
@@ -0,0 +1,19 @@
+// Insert code here to add functionality to your managed object subclass
+
+- (NSString *)firstAlphaCapitalized {
+    return [[self.name substringWithRange:NSMakeRange(0, 1)] uppercaseString];
```
`[self.name capitalizedString]` can be used

soxjke replied:
Thanks for pointing on this method, but for this particular case it's not of more use that current approach - firstAlphaCapitalized is for receiving section name (1 alpha). No matter substring -> uppercase or capitalize -> substring

fuzza commented on DataSources/TECBackgroundContextObjectMutator.m:
```diff
@@ -0,0 +1,123 @@
+
+
+- (void)mutateObjects:(NSArray <NSManagedObject *> *)objects
+withEntityDescription:(NSEntityDescription *)entityDescription
```
I suggest to remove `with` in order to make signature more readable and nice formatted :)

fuzza commented on DataSources/TECBackgroundContextObjectMutator.m:
```diff
@@ -0,0 +1,123 @@
+                block:(TECFetchedResultsControllerContentProviderMutatorArrayChangeBlock)block {
+    NSParameterAssert(objects);
+    NSParameterAssert(block);
+    TECFetchedResultsControllerContentProviderMutatorArrayChangeBlock innerBlock = [block copy];
```
`copy` is not needed anymore

fuzza commented on DataSources/TECBackgroundContextObjectMutator.m:
```diff
@@ -0,0 +1,123 @@
+        fetchRequest.predicate = [NSPredicate predicateWithFormat:@"SELF IN %@", objectIDs];
+        NSArray *fetchedObjects = [context executeFetchRequest:fetchRequest error:&error];
+        for (NSManagedObject *managedObject in fetchedObjects) {
+            [objectsToMutate replaceObjectAtIndex:[objectIDs indexOfObject:managedObject.objectID]
```
move `indexOfObject:` to new line

soxjke replied:
I don't think this affects readability a lot

soxjke replied:
Moreover, i think this doesn't affect readability at all

fuzza commented on DataSources/TECBackgroundContextObjectMutator.m:
```diff
@@ -0,0 +1,123 @@
+        if (innerBlock) {
+            innerBlock(objectRefetched, error);
+            [context save:&error];
+        }
```
we should think if we need both assertion and block check

fuzza commented on DataSources/TECBackgroundContextObjectMutator.m:
```diff
@@ -0,0 +1,123 @@
+- (void)mutateObject:(NSManagedObject *)object
+           withBlock:(TECFetchedResultsControllerContentProviderMutatorObjectChangeBlock)block {
+    NSParameterAssert(object);
+    NSParameterAssert(block);
```
it could be changed to `if(!innerBlock) {return;}`

fuzza commented on DataSources/TECBackgroundContextObjectMutator.m:
```diff
@@ -0,0 +1,123 @@
+}
+
+- (void)deleteObjects:(NSArray <NSManagedObject *> *)objects
+withEntityDescription:(NSEntityDescription *)entityDescription {
```
the same comment as above - we could remove `with`

fuzza commented on DataSources/TECCoreDataSectionModel.h:
```diff
@@ -0,0 +1,18 @@
+@interface TECCoreDataSectionModel : NSObject <TECSectionModelProtocol>
+
+- (instancetype)initWithFetchedResultsSectionInfo:(id<NSFetchedResultsSectionInfo>)info;
+@property (nonatomic, strong, readonly) id <NSFetchedResultsSectionInfo> info;
```
Can we remove section info retain and just pass-through it from FRC

soxjke replied:
No, it has to be adopted to TECSectionModelProtocol

soxjke replied:
However, removing strong reference is a nice point

fuzza commented on DataSources/TECFetchedResultsControllerContentProvider.h:
```diff
@@ -0,0 +1,33 @@
+                 sectionNameKeyPath:(NSString *)sectionNameKeyPath;
+
+- (void)setCurrentRequest:(NSFetchRequest *)currentRequest; // This will trigger reload
+- (NSFetchRequest *)getCopyOfCurrentRequest;
```
we should avoid `get` in names, according to cocoa conventions it is reserved for getting objects by link (e.g `NSError **`)

fuzza commented on DataSources/TECFetchedResultsControllerContentProvider.m:
```diff
@@ -0,0 +1,433 @@
+                                                       atIndexPath:changeset[kTECChangesetIndexPathKey]
+                                                     forChangeType:[changeset[kTECChangesetChangeTypeKey] unsignedIntegerValue]
+                                                      newIndexPath:changeset[kTECChangesetNewIndexPathKey]];
+        }
```
extract to separate method for single changeset processing

fuzza commented on DataSources/TECFetchedResultsControllerContentProvider.m:
```diff
@@ -0,0 +1,433 @@
+    }
+    NSPredicate *filterChangesForDeletedSection =
+    [NSPredicate predicateWithFormat:@"NOT ((%K = %@) AND (%K IN %@))", kTECChangesetKindKey, @(TECChangesetKindRow), [NSString stringWithFormat:@"%@.%@", kTECChangesetIndexPathKey, @"section"], duplicatingIndices];
+    [self.changeSetArray filterUsingPredicate:filterChangesForDeletedSection];
```
break it into small methods, one method per one predicate+filtering array action or smth. like this

fuzza commented on DataSources/TECFetchedResultsControllerContentProvider.m:
```diff
@@ -0,0 +1,433 @@
+                [self.changeSetArray removeObject:changeset];
+            }
+        }
+    }
```
I suggest to break this code to different methods

fuzza commented on DataSources/TECFetchedResultsControllerContentProvider.m:
```diff
@@ -0,0 +1,433 @@
+        if ([changeset[kTECChangesetKindKey] unsignedIntegerValue] == TECChangesetKindRow) {
+            if ([changeset[kTECChangesetChangeTypeKey] unsignedIntegerValue] == NSFetchedResultsChangeMove) {
+                [moves addObject:changeset[kTECChangesetIndexPathKey]];
+            }
```
seems to many indentation levels

fuzza commented on DataSources/TECFetchedResultsControllerContentProvider.m:
```diff
@@ -0,0 +1,433 @@
+    self.fetchedResultsController.fetchRequest.propertiesToGroupBy = currentRequest.propertiesToGroupBy;
+    self.fetchedResultsController.fetchRequest.fetchBatchSize = currentRequest.fetchBatchSize;
+    self.fetchedResultsController.fetchRequest.fetchLimit = currentRequest.fetchLimit;
+    self.fetchedResultsController.fetchRequest.fetchOffset = currentRequest.fetchOffset;
```
what will happen if `currentRequest` will have another Entity. I suggest to set copy of NSFetchRequest to fetched results controller

soxjke replied:
Nice point, but ocassionally, FRC has it's fetchRequest readonly :(

fuzza replied:
@soxjke seems that we have to recreate FRC in this case

soxjke replied:
@fuzza works fine - checked with two sample controllers in app

fuzza commented on DataSources/TECFetchedResultsControllerContentProvider.m:
```diff
@@ -0,0 +1,433 @@
+    self.fetchedResultsController.delegate = self;
+    if (completion) {
+        completion();
+    }
```
I think completion block should be called after presentation adapter reloads view

fuzza commented on DataSources/TECFetchedResultsControllerContentProvider.m:
```diff
@@ -0,0 +1,433 @@
+
+- (void)performBatchUpdatesWithBlock:(TECContentProviderBatchUpdatesBlock)block {
+    if (block) {
+        block(self);
```
why do we set self as parameter? Intention is not clear

fuzza replied:
I think it is not responsibility area of content provider

soxjke committed:
DTCOR-20 - iOS - Add content provider with FRC support

- Code cleanup
- Fixed review notes

soxjke committed:
Merge remote-tracking branch 'origin/develop' into DTCOR-20-add-content-provider-with-frc-support

Conflicts:
	DataSources.xcodeproj/project.pbxproj

soxjke committed:
DTCOR-20 - iOS - Add content provider with FRC support

- Fixed merge debt

soxjke committed:
DTCOR-20 - iOS - Add content provider with FRC support

- Added init tests for FRC

soxjke committed:
Merge remote-tracking branch 'origin/develop' into DTCOR-20-add-content-provider-with-frc-support

Conflicts:
	DataSources.xcodeproj/project.pbxproj

soxjke committed:
DTCOR-20 - iOS - Add content provider with FRC support

- Fixed interfaces for sample app
- Removed redundant registrator layer from UITableView samples

soxjke committed:
DTCOR-20 - iOS - Add content provider with FRC support

- Mote tests in spec

soxjke committed:
DTCOR-20 - iOS - Add content provider with FRC support

- Added changeset/workarounds tests

soxjke committed:
DTCOR-20 - iOS - Add content provider with FRC support

- Completed workaround tests

soxjke committed:
DTCOR-20 - iOS - Add content provider with FRC support

- Completed workaround tests

soxjke committed:
DTCOR-20 - iOS - Add content provider with FRC support

- Added sample with search

soxjke committed:
DTCOR-20 - iOS - Add content provider with FRC support

- Added sample with two entites
